### ISSUE TITLE
[API Proposal]: Allow Leading Unicode Whitespace in NumberStyles

### ISSUE THREAD

**Issue by Megasware128** [dotnet/runtime#86834]
### Background and motivation

Currently, the `NumberStyles` enum allows parsing of leading and trailing white spaces, but only for a subset of Unicode characters. However, in certain scenarios, developers may encounter numbers prefixed by Unicode white-space characters not covered by the current implementation. This issue was first discussed in #86641. This proposal aims to extend the functionality of the `NumberStyles` enum to support leading Unicode white-space characters, thereby offering greater flexibility in handling various text formats.

### API Proposal

```csharp
namespace System.Globalization
{  
    [Flags]
    public enum NumberStyles
    {  
        AllowLeadingWhite = 1,
        AllowTrailingWhite = 2,
        // Proposed addition:
        AllowLeadingUnicodeWhitespace = 2048, // The value can be decided during implementation
        // Rest of the enum values...
    }  
}
```

### API Usage

```csharp
string numberString = "\u2009" + "42"; // Thin space (U+2009) followed by the number "42"
int result;
if (int.TryParse(numberString, NumberStyles.Integer | NumberStyles.AllowLeadingUnicodeWhitespace, CultureInfo.InvariantCulture, out result))
{
    Console.WriteLine(result);  // Outputs: 42
}
else
{
    Console.WriteLine("Failed to parse numberString.");
}
```

### Alternative Designs

An alternative approach could be to update the `AllowLeadingWhite` and `AllowTrailingWhite` flags to include all Unicode white-space characters, but this would be a significant behavioral change and may not be backward compatible.

### Risks

The main risk involves potential breaking changes for existing applications relying on the current `AllowLeadingWhite` and `AllowTrailingWhite` behaviors. It's also important to consider the performance implications of expanding the range of characters that need to be checked during parsing operations.

**Comment by huoyaoyuan** [comment]
A workaround is `numberString.AsSpan().Trim()` (or `TrimStart`/`TrimEnd`), which handles Unicode whitespaces.

**Comment by Megasware128** [comment]
@huoyaoyuan I noticed the issue while parsing a currency. Trimming wouldn't have worked because it had a currency symbol in front of the whitespace. My workaround was this:

```csharp
private static double ParsePrice(string price)
{
    Span<char> chars = stackalloc char[price.Length];

    // Replace non-standard whitespace with regular whitespace
    for (var i = 0; i < price.Length; i++)
    {
        var c = price[i];

        chars[i] = char.IsWhiteSpace(c) ? ' ' : c;
    }

    return double.Parse(chars, NumberStyles.Currency, new CultureInfo("nl-NL"));
}
```

**Comment by danmoseley** [comment]
> currency symbol in front of the whitespace

Then the ws isn't leading - would your proposed API have helped?

**Comment by Megasware128** [comment]
Yes, the proposed `AllowLeadingUnicodeWhitespace` would still have been beneficial. The proposed flag is intended to behave similarly to the existing `AllowLeadingWhite`, but with an extended range of recognized Unicode whitespace characters. Therefore, when combined with `AllowCurrencySymbol`, parsing would still occur if the currency symbol is leading and the Unicode whitespace follows it. The key point here is to expand the range of recognized whitespace characters in various contexts, not just when they are the leading characters.

However, as discussed in the API proposal, there are alternative designs. One of them could be to update the `AllowLeadingWhite` and `AllowTrailingWhite` flags to include all Unicode white-space characters. But this could be a significant behavioral change and may not be backward compatible. The `AllowLeadingUnicodeWhitespace` proposal aims to provide a more flexible solution that can work in tandem with existing flags and behaviors.

**Comment by tannergooding** [comment]
This isn't something we're looking to provide at this time. There are a number of allocation free APIs that allow developers to trivially skip leading whitespace.
